src/webui/api/torrentscontroller.cpp

```diff
@@ -1041,3 +1041,112 @@ void TorrentsController::categoriesAction()
 
     setResult(categories);
 }
+
+void TorrentsController::addTagsAction()
+{
+    checkParams({"hashes", "tags"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+    QJsonArray errors;
+
+    for (const QString &tag : tags) {
+        const QString tagTrimmed {tag.trimmed()};
+        applyToTorrents(hashes, [&tagTrimmed, &errors](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (!torrent->addTag(tagTrimmed) && !torrent->hasTag(tagTrimmed)) {
+                errors.append(QJsonObject {
+                    {"tag", tagTrimmed},
+                    {"hash", static_cast<QString>(torrent->hash())},
+                    {"errorMessage", "Unable to add tag"}
+                });
+            }
+        });
+    }
+
+    setResult(errors);
+}
+
+void TorrentsController::removeTagsAction()
+{
+    checkParams({"hashes"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+    QJsonArray errors;
+
+    for (const QString &tag : tags) {
+        const QString tagTrimmed {tag.trimmed()};
+        applyToTorrents(hashes, [&tagTrimmed, &errors](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (!torrent->removeTag(tagTrimmed) && torrent->hasTag(tagTrimmed)) {
+                errors.append(QJsonObject {
+                    {"tag", tagTrimmed},
+                    {"hash", static_cast<QString>(torrent->hash())},
+                    {"errorMessage", "Unable to remove tag"}
+                });
+            }
+        });
+    }
+
+    if (tags.isEmpty()) {
+        applyToTorrents(hashes, [&errors](BitTorrent::TorrentHandle *const torrent)
+        {
+            torrent->removeAllTags();
+            if (!torrent->tags().isEmpty()) {
```

Comment:
The previous action should not fail. Why do you make this redundant checking and error handling?

Reply:
If we follow this line of reasoning, then we don't need error checking at `removeTag` and `removeAllTags` altogether. They supposed to succeed when tag is valid and exists, do nothing for absent tags, and invalid tags should not be added in the first place. I believe what @Piccirello wants is form of defensive programming, in case internal tags API would change in the future in a way, that would allow these operations to fail. If we won't put any checks and error reporting, web API would fail silently.

Reply:
>I believe what @Piccirello wants is a form of defensive programming, in case internal tags API would change in the future in a way, that would allow these operations to fail.

It's good idea in common case. But you have to have some common sense not to make it absurd.
`removeAllTags` isn't supposed to fail by its meaning. It doesn't even have any error reporting. You're trying to make something up.
As for `removeTag` and others like it, you can insist on "paranoid" error handling like "if a function may report an error, then we have to handle it anyway". But it does not untie your hands to produce stupid logic, trying to perform obviously unsuccessful actions and reporting the same error many times (I mean you should check tag before trying to add/remove it).

Reply:
>  in case internal tags API would change in the future in a way, that would allow these operations to fail. If we won't put any checks and error reporting, web API would fail silently

That's exactly what I was aiming for in this PR.

> removeAllTags isn't supposed to fail by its meaning. It doesn't even have any error reporting. You're trying to make something up.

`TorrentHandle::removeAllTags()` swallows the `false` return from `TorrentHandle::removeTag()`, repeatedly. By checking if the tags list is empty, we can catch this error. The `TorrentHandle` api is pretty lax on error handling.